iml-agent/src/daemon_plugins/postoffice.rs

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::{
+    agent_error::ImlAgentError,
+    env,
+    daemon_plugins::{DaemonPlugin, Output},
+    http_comms::mailbox_client::send,
+};
+use futures::{
+    Future, FutureExt,
+    stream::{StreamExt as _, TryStreamExt},
+};
+use inotify::{Inotify, WatchDescriptor, WatchMask};
+use parking_lot::Mutex;
+use std::{
+    collections::{HashMap, HashSet},
+    pin::Pin,
+    sync::Arc,
+};
+use stream_cancel::{StreamExt, Trigger, Tripwire};
+use tokio::{fs, net::UnixListener};
+use tokio_util::codec::{BytesCodec, FramedRead};
+
+pub struct POWD(pub Option<WatchDescriptor>);
+
+pub struct PostOffice {
+    // individual mailbox socket listeners
+    routes: Arc<Mutex<HashMap<String, Trigger>>>,
+    inotify: Arc<Mutex<Inotify>>,
+    wd: Arc<Mutex<POWD>>,
+}
+
+pub fn create() -> impl DaemonPlugin {
+    PostOffice {
+        wd: Arc::new(Mutex::new(POWD(None))),
+        inotify: Arc::new(Mutex::new(
+            Inotify::init().expect("Failed to initialize inotify"),
+        )),
+        routes: Arc::new(Mutex::new(HashMap::new())),
+    }
+}
+
+/// Return socket address for a given mailbox
+pub fn socket_name(mailbox: &str) -> String {
+    let sock_dir = env::get_var("SOCK_DIR");
+    format!("{}/postman-{}.sock", sock_dir, mailbox)
+}
+
+impl std::fmt::Debug for PostOffice {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "PostOffice {{ {:?} }}", self.routes.lock().keys())
+    }
+}
+
+fn start_route(mailbox: String) -> Trigger {
+    let (trigger, tripwire) = Tripwire::new();
+    let addr = socket_name(&mailbox);
+
+    let rc = async move {
+        // these must live in async because mutable values can't be sent
+        let mut listener = UnixListener::bind(addr.clone())?;
+        let mut incoming = listener.incoming().take_until(tripwire);
+
+        tracing::debug!("Starting Route for {}", mailbox);
+        while let Some(inbound) = incoming.next().await {
+            match inbound {
+                Ok(inbound) => {
+                    let stream = FramedRead::new(inbound, BytesCodec::new())
+                        .map_ok(bytes::BytesMut::freeze)
+                        .map_err(ImlAgentError::Io);
+                    let transfer = send(mailbox.clone(), stream).map(|r| {
+                        if let Err(e) = r {
+                            tracing::error!("Failed to transfer: {}", e);
+                        }
+                    });
+                    tokio::spawn(transfer);
+                }
+                Err(e) => tracing::error!("Failed transfer: {}", e),
+            }
+        }
+        tracing::debug!("Ending Route for {}", mailbox);
+        fs::remove_file(addr).await
+    };
+    tokio::spawn(rc);
+    trigger
+}
+
+fn stop_route(trigger: Trigger) {
```

Comment:
This fn is not needed. `Drop` will be invoked whenever a value goes out of scope.